### PR TITLE
Consider timezones with  `UTC` and `+00:00` to be the same

### LINKED ISSUE #10957: Timezone `UTC` is not considered equal to `+00:00`
### Describe the bug

Running

```sql
select now() - start_timestamp from records limit 5;
```

Results in an error:

```
Cannot coerce arithmetic expression Timestamp(Nanosecond, Some("+00:00")) - Timestamp(Microsecond, Some("UTC")) to valid types
```

As far as I know UTC is alwasy identical to `+00:00` so these two should be considered equivilant?

### To Reproduce

_No response_

### Expected behavior

Timezones which are practically equivalent — e.g. have the same offset and DST semantics should allow subtraction to get an interval.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by marvinlanhenke**
## Which issue does this PR close?

Closes #10957.

## Rationale for this change

Timezone was specified as "+00:00" instead of "UTC". This caused a missmatch in `temporal_coercion`.

### CODE REVIEW

**Comment by marvinlanhenke** [comment]
@alamb PTAL.

Since "+00:00" and "UTC" are the same - I think this is okay? Perhaps, there might me some docs that need some updates to clarify the behavior or documented the supported timezones?

**Comment by alamb** [comment]
I responded on https://github.com/apache/datafusion/issues/10957#issuecomment-2176332944

**Comment by erratic-pattern** [comment]
> Since "+00:00" and "UTC" are the 'same' - is this viable? Or do we have to extend the logic [here](https://github.com/apache/datafusion/blob/main/datafusion/expr/src/type_coercion/binary.rs#L1055) in order to provide both variants (and possibly not introduce a breaking change)? 

+1 on including type coercion logic for "UTC" and "+00:00" and leaving the signature of NowFunc unchanged. This would fix the error descrined in https://github.com/apache/datafusion/issues/10957 more generally without introducing any breaking changes

**Comment by marvinlanhenke** [comment]
...I'm not sure what went wrong with the branch here - I didn't want to close this. I'll reopen (yet...) another PR with the adjusted type coercion. I might need some help here to construct a proper sqllogictest to test this.

**Comment by marvinlanhenke** [comment]
...perhaps someone can help me out on this. I still haven't found a way to create a timestamp with timezone in order to create a proper sqllogictest.

**Comment by alamb** [comment]
> ...perhaps someone can help me out on this. I still haven't found a way to create a timestamp with timezone in order to create a proper sqllogictest.

There are some examples here: https://github.com/apache/datafusion/issues/10602

You can also use arrow_cast (there are some examples here https://datafusion.apache.org/user-guide/sql/scalar_functions.html#arrow-cast)

**Comment by marvinlanhenke** [comment]
I think this should be fixed now. I added some sqllogictest as well to validate the new behavior - thanks @alamb for the pointer on `arrow_cast`. PTAL @alamb @erratic-pattern

File: datafusion/expr/src/type_coercion/binary.rs [hunk 1]
```diff
@@ -1050,12 +1050,14 @@ fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataTyp
...
         (Timestamp(lhs_unit, lhs_tz), Timestamp(rhs_unit, rhs_tz)) => {
             let tz = match (lhs_tz, rhs_tz) {
-                // can't cast across timezones
                 (Some(lhs_tz), Some(rhs_tz)) => {
-                    if lhs_tz != rhs_tz {
-                        return None;
-                    } else {
-                        Some(lhs_tz.clone())
+                    match (lhs_tz.as_ref(), rhs_tz.as_ref()) {
+                        ("UTC", "+00:00") | ("+00:00", "UTC") => Some(lhs_tz.clone()),
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/binary.rs:1055]:
```suggestion
                        // UTC and "+00:00" are the same by definition. Most other timezones
                        // do not have a 1-1 mapping between timezone and an offset from UTC
                        ("UTC", "+00:00") | ("+00:00", "UTC") => Some(lhs_tz.clone()),
```

**Review by alamb — APPROVED** [review]
Thank you @marvinlanhenke  and @samuelcolvin  -- this looks good to me

cc @waitingkuo

**Comment by alamb** [comment]
I added some additional comments and merged up from main.

@erratic-pattern  if you have time, it would be great if you could review this PR as well

**Comment by alamb** [comment]
Thanks everyone!

**Pull request merged**